iroh-net/examples/magic.rs

```diff
@@ -0,0 +1,137 @@
+use std::net::SocketAddr;
+
+use clap::Parser;
+use ed25519_dalek::SigningKey as SecretKey;
+use iroh_net::{
+    defaults::default_derp_map,
+    hp::derp::{DerpMap, UseIpv4, UseIpv6},
+    magic_endpoint::accept_conn,
+    tls::{Keypair, PeerId},
+    MagicEndpoint,
+};
+use tracing::{debug, info};
+use url::Url;
+
+const EXAMPLE_ALPN: &[u8] = b"n0/iroh/examples/magic/0";
+
+#[derive(Debug, Parser)]
+struct Cli {
+    #[clap(short, long)]
+    secret: Option<String>,
+    #[clap(short, long, default_value = "n0/iroh/examples/magic/0")]
+    alpn: String,
+    #[clap(short, long, default_value = "0.0.0.0:0")]
+    bind_address: SocketAddr,
+    #[clap(short, long)]
+    derp_url: Option<Url>,
+    #[clap(subcommand)]
+    command: Command,
+}
+
+#[derive(Debug, Parser)]
+enum Command {
+    Listen,
+    Connect {
+        peer_id: String,
+        addrs: Option<Vec<SocketAddr>>,
+    },
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    tracing_subscriber::fmt::init();
+    let args = Cli::parse();
+    let keypair = match args.secret {
+        None => {
+            let keypair = Keypair::generate();
+            println!("our secret key: {}", fmt_secret(&keypair));
+            keypair
+        }
+        Some(key) => parse_secret(&key)?,
+    };
+
+    let derp_map = match args.derp_url {
+        None => default_derp_map(),
+        Some(url) => {
+            // TODO: This should be done by the DERP client.
+            let derp_port = match url.port() {
+                Some(port) => port,
+                None => match url.scheme() {
+                    "http" => 80,
+                    "https" => 443,
+                    _ => anyhow::bail!(
+                        "Invalid scheme in DERP URL, only http: and https: schemes are supported."
+                    ),
```

Comment:
I have seen this logic in your other PR, maybe we can add a helper somewhere to keep this consistent?

Reply:
This would be obsoleted with https://github.com/n0-computer/iroh/pull/1143 I think because then we don't have to the port parsing here and just pass the URL through